spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.puffin.Blob;
+import org.apache.iceberg.puffin.Puffin;
+import org.apache.iceberg.puffin.PuffinWriter;
+import org.apache.iceberg.puffin.StandardBlobTypes;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.function.Function2;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.internal.SQLConf;
+import scala.Tuple2;
+
+public class NDVSketchGenerator {
+
+  private NDVSketchGenerator() {}
+
+  static StatisticsFile generateNDV(
+      SparkSession spark, Table table, long snapshotId, String... columnsToBeAnalyzed)
+      throws IOException {
+    List<String> columnList;
+    if (columnsToBeAnalyzed == null || columnsToBeAnalyzed.length == 0) {
+      columnList =
+          table.schema().columns().stream()
+              .map(Types.NestedField::name)
+              .collect(Collectors.toList());
+    } else {
+      columnList = Lists.newArrayList(columnsToBeAnalyzed);
+    }
+    Iterator<Tuple2<String, ThetaSketchJavaSerializable>> tuple2Iterator =
+        computeNDVSketches(spark, table.name(), snapshotId, columnList);
+    Map<String, ThetaSketchJavaSerializable> sketchMap = Maps.newHashMap();
+
+    tuple2Iterator.forEachRemaining(tuple -> sketchMap.put(tuple._1, tuple._2));
+    return writeToPuffin(table, columnList, sketchMap);
+  }
+
+  private static StatisticsFile writeToPuffin(
+      Table table, List<String> columns, Map<String, ThetaSketchJavaSerializable> sketchMap)
+      throws IOException {
+    TableOperations operations = ((HasTableOperations) table).operations();
+    FileIO fileIO = operations.io();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = fileIO.newOutputFile(path);
+    try (PuffinWriter writer =
+        Puffin.write(outputFile).createdBy("Iceberg Analyze action").build()) {
+      for (String columnName : columns) {
+        writer.add(
+            new Blob(
+                StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                ImmutableList.of(table.schema().findField(columnName).fieldId()),
+                table.currentSnapshot().snapshotId(),
+                table.currentSnapshot().sequenceNumber(),
+                ByteBuffer.wrap(sketchMap.get(columnName).getSketch().toByteArray()),
```

Comment:
BTW it would be good to have a cross-engine compatibility test to ensure the value we write here can indeed be used correctly by other engines. for trino, you can use https://java.testcontainers.org/modules/databases/trino/ 

Trino already has such tests, but that doesn't cover Iceberg Spark features that are **being** implemented.